File: app/src/main/java/com/example/waveplayer/FragmentEditPlaylist.java
package com.example.waveplayer;

import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;

import java.util.ArrayList;
import java.util.List;

public class FragmentEditPlaylist extends Fragment {

    ActivityMain activityMain;

    View view;

    BroadcastReceiverOnServiceConnected broadcastReceiverOnServiceConnected;

    OnClickListenerFragmentEditPlaylistButtonSelectSongs
            onClickListenerFragmentEditPlaylistButtonSelectSongs;

    OnClickListenerFABFragmentEditPlaylist onClickListenerFABFragmentEditPlaylist;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_edit_playlist, container, false);
        onClickListenerFragmentEditPlaylistButtonSelectSongs =
                new OnClickListenerFragmentEditPlaylistButtonSelectSongs(this);
        activityMain = ((ActivityMain) getActivity());
        if (activityMain != null) {
            activityMain.setActionBarTitle(getResources().getString(R.string.edit_playlist));
        }
        updateFAB(view);
        view.findViewById(R.id.buttonEditSongs).setOnClickListener(
                onClickListenerFragmentEditPlaylistButtonSelectSongs);
        setUpBroadcastReceiverServiceConnected(view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    private void updateFAB(View view) {
        activityMain.setFabImage(R.drawable.ic_check_black_24dp);
        activityMain.setFABText(R.string.fab_save);
        activityMain.showFab(true);
        final EditText finalEditTextPlaylistName = view.findViewById(R.id.editTextPlaylistName);
        ArrayList<AudioUri> playlistSongs = new ArrayList<>();
        RandomPlaylist userPickedPlaylist = activityMain.getUserPickedPlaylist();
        // userPickedPlaylist is null when user is making a new playlist
        if (userPickedPlaylist != null) {
            List<AudioUri> userPickedSongs = activityMain.getUserPickedSongs();
            // userPickedSongs.isEmpty() when the user is editing a playlist
            if (userPickedSongs.isEmpty()) {
                userPickedSongs.addAll(userPickedPlaylist.getAudioUris());
            }
            playlistSongs = new ArrayList<>(userPickedPlaylist.getAudioUris());
            finalEditTextPlaylistName.setText(userPickedPlaylist.getName());
        }
        final ArrayList<AudioUri> finalPlaylistSongs = playlistSongs;
        onClickListenerFABFragmentEditPlaylist = new OnClickListenerFABFragmentEditPlaylist(
                userPickedPlaylist, this, finalPlaylistSongs,
                finalEditTextPlaylistName);
        activityMain.setFabOnClickListener(onClickListenerFABFragmentEditPlaylist);
    }

    void popBackStackAndHideKeyboard(View view) {
        NavController navController = NavHostFragment.findNavController(
                FragmentEditPlaylist.this);
        navController.popBackStack();
        InputMethodManager imm = (InputMethodManager) activityMain.getSystemService(
                AppCompatActivity.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

    private void setUpBroadcastReceiverServiceConnected(final View view) {
        IntentFilter filterComplete = new IntentFilter();
        filterComplete.addCategory(Intent.CATEGORY_DEFAULT);
        filterComplete.addAction(activityMain.getResources().getString(
                R.string.broadcast_receiver_action_service_connected));
        broadcastReceiverOnServiceConnected = new BroadcastReceiverOnServiceConnected() {
            @Override
            public void onReceive(Context context, Intent intent) {
                updateFAB(view);
            }
        };
        activityMain.registerReceiver(broadcastReceiverOnServiceConnected, filterComplete);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        activityMain.unregisterReceiver(broadcastReceiverOnServiceConnected);
        view.findViewById(R.id.buttonEditSongs).setOnClickListener(null);
        onClickListenerFragmentEditPlaylistButtonSelectSongs = null;
        onClickListenerFABFragmentEditPlaylist = null;
        activityMain = null;
        view = null;
    }

}

File: app/src/main/java/io/fourthFinger/pinkyPlayer/fragments/FragmentEditPlaylist.kt
package io.fourthFinger.pinkyPlayer.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.NavHostFragment
import io.fourthFinger.pinkyPlayer.KeyboardUtil
import io.fourthFinger.pinkyPlayer.R
import io.fourthFinger.pinkyPlayer.activity_main.ViewModelActivityMain
import io.fourthFinger.pinkyPlayer.databinding.FragmentEditPlaylistBinding

class FragmentEditPlaylist : Fragment() {

    private var _binding: FragmentEditPlaylistBinding? = null
    private val binding get() = _binding!!

    private val viewModelActivityMain by activityViewModels<ViewModelActivityMain> {
        ViewModelActivityMain.Factory
    }

    private val viewModelEditPlaylist by activityViewModels<ViewModelFragmentEditPlaylist> {
        ViewModelFragmentEditPlaylist.Factory
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        viewModelActivityMain.setActionBarTitle(resources.getString(R.string.edit_playlist))
        _binding = FragmentEditPlaylistBinding.inflate(
            inflater,
            container,
            false
        )
        return binding.root
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?
    ) {
        super.onViewCreated(view, savedInstanceState)
        binding.buttonEditSongs.setOnClickListener {
            viewModelEditPlaylist.selectSongsClicked(
                NavHostFragment.findNavController(this)
            )
        }
    }

    override fun onResume() {
        super.onResume()
        updateFAB()
    }

    private fun updateFAB() {
        viewModelActivityMain.setFabImage(R.drawable.ic_check_black_24dp)
        viewModelActivityMain.setFABText(R.string.fab_save)
        viewModelActivityMain.showFab(true)
        val editTextPlaylistName: EditText = binding.editTextPlaylistName
        // TODO why the conditional?
        if (editTextPlaylistName.text.isEmpty()) {
            editTextPlaylistName.setText(
                viewModelActivityMain.currentContextPlaylist?.getName()
            )
        }
        viewModelActivityMain.setFabOnClickListener { view: View ->
            viewModelEditPlaylist.editPlaylistFabClicked(
                this,
                editTextPlaylistName.text.toString()
            )
            cleanUp(view)
        }
    }

    private fun cleanUp(view: View) {
        KeyboardUtil.hideKeyboard(view)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModelActivityMain.setFabOnClickListener(null)
    }

}